File: apps/api/services/director/analysis/reconcile_diarization.py
"""Diarization reconciliation for chunked analysis (Phase 12)."""
from __future__ import annotations

import math
from typing import Any

from services.director.analysis.plan_chunks import ChunkPlan


SIMILARITY_THRESHOLD = 0.82


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(y * y for y in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def _segments_in_window(
    segments: list[dict[str, Any]],
    start: float,
    end: float,
) -> list[dict[str, Any]]:
    return [
        s
        for s in segments
        if float(s.get("end", 0)) > start and float(s.get("start", 0)) < end
    ]


def reconcile_diarization(
    chunk_outputs: list[tuple[ChunkPlan, list[dict[str, Any]], dict[str, list[float]] | None]],
) -> list[dict[str, Any]]:
    """
    Reconcile local speaker IDs across chunks into global, file-consistent IDs.

    chunk_outputs: list of (chunk_plan, segments, optional speaker_embeddings)
    """
    if not chunk_outputs:
        return []
    if len(chunk_outputs) == 1:
        return list(chunk_outputs[0][1])

    global_map: dict[str, str] = {}
    next_global = 0
    chunk_local_to_global: list[dict[str, str]] = []

    def assign_global(local_key: str) -> str:
        nonlocal next_global
        if local_key in global_map:
            return global_map[local_key]
        gid = f"G{next_global}"
        next_global += 1
        global_map[local_key] = gid
        return gid

    for i, (chunk, segments, speaker_embeddings) in enumerate(chunk_outputs):
        local_map: dict[str, str] = {}
        speaker_embeddings = speaker_embeddings or {}

        if i == 0:
            for seg in segments:
                sid = str(seg.get("speakerId", "A"))
                key = f"{chunk.chunk_index}:{sid}"
                local_map[sid] = assign_global(key)
            chunk_local_to_global.append(local_map)
            continue

        prev_chunk, prev_segments, prev_embeddings = chunk_outputs[i - 1]
        prev_map = chunk_local_to_global[i - 1]
        prev_embeddings = prev_embeddings or {}

        overlap_start = max(prev_chunk.window_start, chunk.window_start)
        overlap_end = min(prev_chunk.window_end, chunk.window_end)

        prev_overlap = _segments_in_window(prev_segments, overlap_start, overlap_end)
        curr_overlap = _segments_in_window(segments, overlap_start, overlap_end)

        for curr_seg in curr_overlap:
            curr_id = str(curr_seg.get("speakerId", "A"))
            best_global: str | None = None
            best_score = SIMILARITY_THRESHOLD
            curr_emb = speaker_embeddings.get(curr_id)

            for prev_seg in prev_overlap:
                prev_id = str(prev_seg.get("speakerId", "A"))
                prev_emb = prev_embeddings.get(prev_id)
                if curr_emb and prev_emb:
                    score = _cosine_similarity(curr_emb, prev_emb)
                else:
                    overlap = min(
                        float(curr_seg.get("end", 0)),
                        float(prev_seg.get("end", 0)),
                    ) - max(
                        float(curr_seg.get("start", 0)),
                        float(prev_seg.get("start", 0)),
                    )
                    score = 0.85 if overlap > 0 and curr_id == prev_id else 0.0

                if score >= best_score:
                    best_score = score
                    best_global = prev_map.get(prev_id)

            if best_global:
                local_map[curr_id] = best_global

        for seg in segments:
            sid = str(seg.get("speakerId", "A"))
            if sid not in local_map:
                key = f"{chunk.chunk_index}:{sid}"
                local_map[sid] = assign_global(key)

        chunk_local_to_global.append(local_map)

    reconciled: list[dict[str, Any]] = []
    for i, (chunk, segments, _) in enumerate(chunk_outputs):
        local_map = chunk_local_to_global[i]
        for seg in segments:
            start = float(seg.get("start", 0))
            end = float(seg.get("end", start))
            mid = (start + end) / 2
            if mid < chunk.core_start or mid > chunk.core_end:
                continue
            sid = str(seg.get("speakerId", "A"))
            reconciled.append({**seg, "speakerId": local_map.get(sid, sid)})

    return sorted(reconciled, key=lambda s: float(s.get("start", 0)))

File: remotion-service/src/lib/analysis/reconcileDiarization.ts
import { type ChunkPlan } from "./planChunks";

export interface SpeakerSegment {
  start: number;
  end: number;
  confidence: number;
  speakerId: string;
  confidenceSource?: string;
  embedding?: number[];
}

export interface DiarizationChunkOutput {
  chunk: ChunkPlan;
  segments: SpeakerSegment[];
  /** Optional centroid per local speaker id for cross-chunk matching. */
  speakerEmbeddings?: Record<string, number[]>;
}

const SIMILARITY_THRESHOLD = 0.82;

function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length === 0 || b.length === 0 || a.length !== b.length) {
    return 0;
  }
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i += 1) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  if (normA === 0 || normB === 0) {
    return 0;
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

function segmentsInWindow(
  segments: SpeakerSegment[],
  start: number,
  end: number,
): SpeakerSegment[] {
  return segments.filter((s) => s.end > start && s.start < end);
}

/**
 * Reconcile local speaker IDs across chunks into global, file-consistent IDs.
 * Uses embedding cosine similarity when available; falls back to temporal overlap
 * at chunk boundaries.
 */
export function reconcileDiarization(
  chunkOutputs: DiarizationChunkOutput[],
): SpeakerSegment[] {
  if (chunkOutputs.length === 0) {
    return [];
  }
  if (chunkOutputs.length === 1) {
    return [...chunkOutputs[0].segments];
  }

  const globalMap = new Map<string, string>();
  let nextGlobalId = 0;

  const assignGlobal = (localKey: string): string => {
    const existing = globalMap.get(localKey);
    if (existing) {
      return existing;
    }
    const id = `G${nextGlobalId}`;
    nextGlobalId += 1;
    globalMap.set(localKey, id);
    return id;
  };

  const chunkLocalToGlobal: Array<Map<string, string>> = [];

  for (let i = 0; i < chunkOutputs.length; i += 1) {
    const { chunk, segments, speakerEmbeddings } = chunkOutputs[i];
    const localMap = new Map<string, string>();

    if (i === 0) {
      for (const seg of segments) {
        const key = `${chunk.chunkIndex}:${seg.speakerId}`;
        localMap.set(seg.speakerId, assignGlobal(key));
      }
      chunkLocalToGlobal.push(localMap);
      continue;
    }

    const prev = chunkOutputs[i - 1];
    const prevMap = chunkLocalToGlobal[i - 1];
    const overlapStart = Math.max(prev.chunk.windowStart, chunk.windowStart);
    const overlapEnd = Math.min(prev.chunk.windowEnd, chunk.windowEnd);

    const prevOverlap = segmentsInWindow(prev.segments, overlapStart, overlapEnd);
    const currOverlap = segmentsInWindow(segments, overlapStart, overlapEnd);

    const matchedLocal = new Set<string>();

    for (const currSeg of currOverlap) {
      let bestGlobal: string | null = null;
      let bestScore = SIMILARITY_THRESHOLD;

      const currEmb = speakerEmbeddings?.[currSeg.speakerId];
      for (const prevSeg of prevOverlap) {
        const prevEmb = prev.speakerEmbeddings?.[prevSeg.speakerId];
        let score = 0;
        if (currEmb && prevEmb) {
          score = cosineSimilarity(currEmb, prevEmb);
        } else {
          const overlap =
            Math.min(currSeg.end, prevSeg.end) - Math.max(currSeg.start, prevSeg.start);
          score = overlap > 0 && currSeg.speakerId === prevSeg.speakerId ? 0.85 : 0;
        }
        if (score >= bestScore) {
          bestScore = score;
          bestGlobal = prevMap.get(prevSeg.speakerId) ?? null;
        }
      }

      if (bestGlobal) {
        localMap.set(currSeg.speakerId, bestGlobal);
        matchedLocal.add(currSeg.speakerId);
      }
    }

    for (const seg of segments) {
      if (!localMap.has(seg.speakerId)) {
        const key = `${chunk.chunkIndex}:${seg.speakerId}`;
        localMap.set(seg.speakerId, assignGlobal(key));
      }
    }

    chunkLocalToGlobal.push(localMap);
  }

  const reconciled: SpeakerSegment[] = [];
  for (let i = 0; i < chunkOutputs.length; i += 1) {
    const { chunk, segments } = chunkOutputs[i];
    const localMap = chunkLocalToGlobal[i];
    for (const seg of segments) {
      const mid = (seg.start + seg.end) / 2;
      if (mid < chunk.coreStart || mid > chunk.coreEnd) {
        continue;
      }
      reconciled.push({
        ...seg,
        speakerId: localMap.get(seg.speakerId) ?? seg.speakerId,
      });
    }
  }

  return reconciled.sort((a, b) => a.start - b.start);
}
